src/merkle_tree/constraints.rs

```diff
@@ -12,100 +12,71 @@ use ark_std::borrow::Borrow;
-        P::TwoToOneHash,
-        ConstraintF,
-        OutputVar = Self::InnerDigest,
+pub trait ConfigGadget<ConstraintF: Field>: Config
+where
+    Self::LeafHash: CRHSchemeGadget<ConstraintF, InputVar = Self::LeafVar>,
+    Self::TwoToOneHash: TwoToOneCRHSchemeGadget<ConstraintF>,
+{
+    type LeafVar: Debug + ?Sized;
+
+    type LeafToInnerVarConverter: DigestVarConverter<
+        LeafDigestVar<Self, ConstraintF>,
+        TwoToOneInputVar<Self, ConstraintF>,
     >;
 }
 
-type LeafParam<PG, P, ConstraintF> =
-    <<PG as ConfigGadget<P, ConstraintF>>::LeafHash as CRHSchemeGadget<
-        <P as Config>::LeafHash,
-        ConstraintF,
-    >>::ParametersVar;
-type TwoToOneParam<PG, P, ConstraintF> =
-    <<PG as ConfigGadget<P, ConstraintF>>::TwoToOneHash as TwoToOneCRHSchemeGadget<
-        <P as Config>::TwoToOneHash,
-        ConstraintF,
-    >>::ParametersVar;
+pub type TwoToOneInputVar<C, CF> =
+    <<C as Config>::TwoToOneHash as TwoToOneCRHSchemeGadget<CF>>::InputVar;
+pub type LeafDigestVar<C, CF> = <<C as Config>::LeafHash as CRHSchemeGadget<CF>>::OutputVar;
+pub type TwoToOneDigestVar<C, CF> =
+    <<C as Config>::TwoToOneHash as TwoToOneCRHSchemeGadget<CF>>::OutputVar;
+
+pub type LeafParamsVar<CG, ConstraintF> =
+    <<CG as Config>::LeafHash as CRHSchemeGadget<ConstraintF>>::ParametersVar;
+
+pub type TwoToOneParamsVar<CG, ConstraintF> =
+    <<CG as Config>::TwoToOneHash as TwoToOneCRHSchemeGadget<ConstraintF>>::ParametersVar;
 
 /// Represents a merkle tree path gadget.
 #[derive(Debug, Derivative)]
-#[derivative(Clone(bound = "P: Config, ConstraintF: Field, PG: ConfigGadget<P, ConstraintF>"))]
-pub struct PathVar<P: Config, ConstraintF: Field, PG: ConfigGadget<P, ConstraintF>> {
+#[derivative(Clone(bound = "P: ConfigGadget<ConstraintF>, ConstraintF: Field"))]
+pub struct PathVar<P: ConfigGadget<ConstraintF>, ConstraintF: Field>
+where
+    P::LeafHash: CRHSchemeGadget<ConstraintF, InputVar = P::LeafVar>,
+    P::TwoToOneHash: TwoToOneCRHSchemeGadget<ConstraintF>,
+{
     /// `path[i]` is 0 (false) iff ith non-leaf node from top to bottom is left.
     path: Vec<Boolean<ConstraintF>>,
     /// `auth_path[i]` is the entry of sibling of ith non-leaf node from top to bottom.
-    auth_path: Vec<PG::InnerDigest>,
+    auth_path: Vec<TwoToOneDigestVar<P, ConstraintF>>,
     /// The sibling of leaf.
-    leaf_sibling: PG::LeafDigest,
+    leaf_sibling: LeafDigestVar<P, ConstraintF>,
     /// Is this leaf the right child?
     leaf_is_right_child: Boolean<ConstraintF>,
 }
 
-impl<P: Config, ConstraintF: Field, PG: ConfigGadget<P, ConstraintF>> AllocVar<Path<P>, ConstraintF>
-    for PathVar<P, ConstraintF, PG>
+impl<P: ConfigGadget<ConstraintF>, ConstraintF: Field> AllocVar<Path<P>, ConstraintF>
+    for PathVar<P, ConstraintF>
 where
-    P: Config,
-    ConstraintF: Field,
+    P::LeafHash: CRHSchemeGadget<ConstraintF, InputVar = P::LeafVar>,
+    P::TwoToOneHash: TwoToOneCRHSchemeGadget<ConstraintF>,
```

Comment:
I think the bounds are probably not needed? The bound is already present in `ConfigGadget` trait. Any config without such bound will not implement `ConfigGadget` correctly, no?

Reply:
Unfortunately the bounds don't propagate  =(

There's [an RFC to fix that](https://github.com/rust-lang/rust/issues/44491), but it's not implemented yet. Once it's implemented, I think the size of arkworks libraries will shrink by a large margin lol

Reply:
oh no... hope that RFC will be stable soon